python/gs/jit/optimize.py

```diff
@@ -1,2 +1,119 @@
-def optimize(ir):
-    pass
\ No newline at end of file
+import operator
+from .passes import dce
+import torch.fx as fx
+
+
+def merge_relabel_and_all_indices(gm: fx.GraphModule) -> fx.GraphModule:
+    merge_dir = {}
+
+    # scan
+    for node in gm.graph.nodes:
+        if node.target == 'all_indices':
+            # node.args[0] is the parent of node
+            if node.args[0] not in merge_dir:
+                merge_dir[node.args[0]] = [node]
+            else:
+                merge_dir[node.args[0]].append(node)
+
+        if node.target == 'relabel':
+            # node.args[0] is the parent of node
+            if node.args[0] not in merge_dir:
+                merge_dir[node.args[0]] = [node]
+            else:
+                merge_dir[node.args[0]].append(node)
+
+    # begin merge
+    for key, value in merge_dir.items():
+
+        if len(value) < 2:
+            continue
+
+        if len(value) > 2:
+            print("{} has more than two children: {}, something wrong?".format(
+                key, value))
+
+        all_indices_node = None
+        relabel_node = None
+        all_indices_is_before = None
+
+        if value[0].target == 'all_indices' and value[1].target == 'relabel':
+            all_indices_node = value[0]
+            relabel_node = value[1]
+            all_indices_is_before = True
+        elif value[0].target == 'relabel' and value[1].target == 'all_indices':
+            all_indices_node = value[1]
+            relabel_node = value[0]
+            all_indices_is_before = False
+        else:
+            raise ValueError
+
+        if all_indices_is_before:
+            with gm.graph.inserting_before(all_indices_node):
```

Comment:
Can we consistently copy the relabel_node no matter whether all_indices is before or after the relabel node? In this case, we only need to maintain the code of one branch, which will reduces the cost to maintain this fusion if later we change the interface of the relabel.